File: Streamdeck +/edit streamdeck.py
import os
import io
from PIL import Image, ImageDraw, ImageFont
from StreamDeck.DeviceManager import DeviceManager
from StreamDeck.ImageHelpers import PILHelper

# Folder location of image assets used by this example.
ASSETS_PATH = os.path.join(os.path.dirname(__file__), "Assets")

def set_key_image(deck, key, image_path, text=None):
    # Open the image file
    icon = Image.open(image_path)

    # Convert the image to RGB mode if it's not already
    if icon.mode != 'RGB':
        icon = icon.convert('RGB')

    # Resize the image to fit the Stream Deck key
    image = PILHelper.create_scaled_key_image(deck, icon)
    
        # Draw text on the image if provided
    if text:
        draw = ImageDraw.Draw(image)
        font_path = os.path.join(ASSETS_PATH, "AovelSansRounded-rdDL.ttf")  # Ensure you have a font file in the Assets folder
        font = ImageFont.truetype(font_path, 14)
        text_bbox = draw.textbbox((0, 0), text, font=font)
        text_width = text_bbox[2] - text_bbox[0]
        text_height = text_bbox[3] - text_bbox[1]
        text_position = ((image.width - text_width) // 2, (image.height - text_height) // 2)
        draw.text(text_position, text, font=font, fill="white")

    # Convert the image to the native format of the Stream Deck
    native_image = PILHelper.to_native_key_format(deck, image)

    # Set the image on the specified key
    deck.set_key_image(key, native_image)

def set_touchscreen_image(deck, image_path):
    # Open the image file
    tscreen = Image.open(image_path)

    # Convert the image to RGB mode if it's not already
    if tscreen.mode != 'RGB':
        tscreen = tscreen.convert('RGB')

    # Resize the image to fit the Stream Deck touchscreen
    image = tscreen.resize((800, 100), Image.LANCZOS)

    # Convert the image to the native format of the Stream Deck touchscreen
    img_byte_arr = io.BytesIO()
    image.save(img_byte_arr, format='JPEG')
    native_image = img_byte_arr.getvalue()

    # Set the image on the touchscreen
    deck.set_touchscreen_image(native_image, 0, 0, 800, 100)

if __name__ == "__main__":
    # Enumerate all connected Stream Deck devices
    streamdecks = DeviceManager().enumerate()

    if len(streamdecks) == 0:
        print("No Stream Deck devices found.")
        exit()

    # Use the first Stream Deck device
    deck = streamdecks[0]

    # Open the Stream Deck device
    deck.open()

    # Reset the Stream Deck device
    deck.reset()

    # Set the brightness of the Stream Deck
    deck.set_brightness(60)

    # Path to the image files
    IMMORTAL = os.path.join(ASSETS_PATH, "immortal.png")
    touch = os.path.join(ASSETS_PATH, "long.png")
    whitesquare = os.path.join(ASSETS_PATH, "whitesquare.jpg")

    # Set the image on the specified key
    set_key_image(deck, 0, IMMORTAL)
    set_key_image(deck, 1, whitesquare, text="Speed Camera")

    # Set the image on the touchscreen
    set_touchscreen_image(deck, touch)

    # Keep the script running to maintain the image on the key
    input("Press Enter to exit...")

    # Close the Stream Deck device
    deck.close()

File: Streamdeck +/editstreamdeck.js
const HID = require('node-hid');
const sharp = require('sharp');
const path = require('path');
const fs = require('fs');

// Folder location of image assets used by this example.
const ASSETS_PATH = path.join(__dirname, 'Assets');

async function setKeyImage(deck, key, imagePath, text = null) {
    try {
        // Open the image file
        let icon = sharp(imagePath);

        // Convert the image to RGB mode if it's not already
        const metadata = await icon.metadata();
        if (metadata.channels !== 3) {
            icon = icon.toColourspace('rgb');
        }

        // Resize the image to fit the Stream Deck key
        const resizedImage = await icon.resize(deck.keyImageFormat().width, deck.keyImageFormat().height).toBuffer();

        // Draw text on the image if provided
        if (text) {
            const fontPath = path.join(ASSETS_PATH, 'AovelSansRounded-rdDL.ttf');  // Ensure you have a font file in the Assets folder
            const svgText = `
                <svg width="${deck.keyImageFormat().width}" height="${deck.keyImageFormat().height}">
                    <text x="50%" y="90%" font-family="${fontPath}" font-size="14" fill="white" text-anchor="middle">${text}</text>
                </svg>
            `;
            const textImage = await sharp(Buffer.from(svgText)).toBuffer();
            const combinedImage = await sharp(resizedImage).composite([{ input: textImage, gravity: 'south' }]).toBuffer();
            deck.setKeyImage(key, combinedImage);
        } else {
            deck.setKeyImage(key, resizedImage);
        }
    } catch (error) {
        console.error(`Error setting key image: ${error}`);
    }
}

// Example usage
async function main() {
    const devices = HID.devices();
    const streamDecks = devices.filter(device => device.vendorId === 4057 && device.productId === 132); // Replace with actual vendorId and productId

    if (streamDecks.length === 0) {
        console.log('No Stream Deck devices found.');
        return;
    }

    const deck = new HID.HID(streamDecks[0].path);
    await setKeyImage(deck, 0, path.join(ASSETS_PATH, 'example.png'), 'Hello World');
    deck.close();
}

main();